### ISSUE TITLE
Embedding library in the toplevel

### ISSUE THREAD

**Issue by picojulien** [ocsigen/js_of_ocaml#585]
Hi,
I would like to build a toplevel with a library of mine available. 

I used to do it by reusing the toplevel example and just adding my_module.cmo  to the mktop line and also "-export-unit my_module".

Now I can't do "-export-unit my_module" anymore as it complains that it can't find package my_module.
using  "-export-unit my_module.cmi -export-unit my_module.cmo", It compiles but I get the error "Reference to undefined global `My_module' " when I try to use it (My_module.test () ).

I attached an example with minimal modifications in the files from js_of_ocaml 2.8.4.
I'm using  js_of_ocaml 2.8.4 with ocaml  4.04.1.
[toplevel.tar.gz](https://github.com/ocsigen/js_of_ocaml/files/1035802/toplevel.tar.gz)

And here is a version which works with  js_of_ocaml 2.6 and ocaml 4.02.3
[toplevel.tar.gz](https://github.com/ocsigen/js_of_ocaml/files/1036357/toplevel.tar.gz)

**Comment by timothee-haudebourg** [comment]
Same problem here.

**Comment by hhugo** [comment]
Sorry for the delay. Can you try with https://github.com/ocsigen/js_of_ocaml/pull/592

**Comment by timothee-haudebourg** [comment]
Hi,
I still have the same problem even with this fix.
Note that I'm trying to load an entire library .cma. It's seems to work since I can open any module of the library in the toplevel. But when I try to use it I get the same error ```Reference to undefined global `Module'``` even if it is opened.

The cmis are loaded since I get an error ```uncaught exception: 0,,/cmis/module.cmi : file already exists``` from the browser when I try to export the corresponding cmis files with ```-export-unit module.cmi```.

**Comment by timothee-haudebourg** [comment]
I should add that I can use any constructor defined I the library/module:
```
                                        OCaml version 4.04.1
     Compiled with Js_of_ocaml version 2.8.4+db4f967

# Module.Cons "foo";;
- : Module.type = Module.Cons "foo"
```
The only problem is when I try to use a function defined in the module:
```
# Module.function 0;;
Characters -1--1:
Error: Reference to undefined global `Module'
```

**Comment by hhugo** [comment]
Is there some code I can look at ? Can you send me the command you run and the logs, if any ?

**Comment by picojulien** [comment]
I think I attached a minimal example with m'y bug report didn't I ?

Le 14 juin 2017 12:15:31 UTC+02:00, hhugo <notifications@github.com> a écrit :
>Is there some code I can look at ? Can you send me the command you run
>and the logs, if any ? 
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/ocsigen/js_of_ocaml/issues/585#issuecomment-308388255

-- 
Envoyé de mon téléphone Android avec K-9 Mail. Excusez la brièveté.

**Comment by timothee-haudebourg** [comment]
> I think I attached a minimal example with m'y bug report didn't I ? 
So here is an other minimal example that doesn't work: [minimal_example.tar.gz](https://github.com/ocsigen/js_of_ocaml/files/1074458/minimal_example.tar.gz)

It includes a module ```Module``` that have the same signature as in my previous toy example. Just do a ```make``` in the root directory it should work. In the resulting toplevel, from your browser, you can ```open Module```, you can do ```Module.Cons "foo"```, but you can't call ```Module.my_function 42```.

**Comment by hhugo** [comment]
@picojulien, yes, and I think https://github.com/ocsigen/js_of_ocaml/pull/592 is the right fix for you. I was able to make your example work.

@timothee-haudebourg, I'll look at this later this week

**Comment by hhugo** [comment]
@timothee-haudebourg, your not linking your library in the bytecode program.
The following change (together with my previous patch) is enough to make you example work. 
```
$(TARGET).js: $(TARGET).cma ../lib/library.cma
	$(JS_MKTOP) $^ -o $(TARGET).byte
```

**Comment by timothee-haudebourg** [comment]
So it was my fault, sorry for the inconvenience. And thanks :)

**Comment by hhugo** [comment]
Thank you for the report.

**Comment by hhugo** [comment]
Closing since the fix was included in the latest version of js_of_ocaml (3.0). Reopen if necessary
